File: Tests/iOS Tests/Tests/PNPublishWithMobilePayloadTests.m
//
//  PNPublishWithMobilePayloadTests.m
//  PubNub Tests
//
//  Created by Vadim Osovets on 6/22/15.
//
//

#import <PubNub/PubNub.h>

#import "PNBasicClientTestCase.h"

#import "NSString+PNTest.h"

@interface PNPublishWithMobilePayloadTests : PNBasicClientTestCase

@end

@implementation PNPublishWithMobilePayloadTests

- (BOOL)isRecording{
    return NO;
}

- (NSString *)publishChannelName {
    return @"02290046-2F36-43DD-97F0-2F51D925451A";
}

- (void)testSimplePublishSimpleMobilePushPayload {
    
    NSDictionary *payload = @{@"aps" :
                                  @{@"alert" : @"You got your emails.@",
                                    @"badge" : @(9),
                                    @"sound" : @"bingbong.aiff"},
                              @"acme 1" : @(42)};
    
    [self performVerifiedPublish:@"test"
                       onChannel:[self publishChannelName]
               mobilePushPayload:payload
                  storeInHistory:YES
                      compressed:YES
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @14587651644081128);
                  }];
}

- (void)testSimplePublishNilMobilePushPayload {
    [self performVerifiedPublish:@"test"
                       onChannel:[self publishChannelName]
               mobilePushPayload:nil
                  storeInHistory:YES
                      compressed:YES
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @14587651642570398);
                  }];
}

- (void)testPublishMobilePayloadNotStoreInHistory {
    
    NSDictionary *payload = @{@"aps" :
                                  @{@"alert" : @"You got your emails.@",
                                    @"badge" : @(9),
                                    @"sound" : @"bingbong.aiff"},
                              @"acme 1" : @(42)};
    
    [self performVerifiedPublish:@"test"
                       onChannel:[self publishChannelName]
               mobilePushPayload:payload
                  storeInHistory:NO
                      compressed:YES
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @14587651639789412);
                  }];
}

- (void)testPublishMobilePayloadNotStoreInHistoryNotCompressed {
    
    NSDictionary *payload = @{@"aps" :
                                  @{@"alert" : @"You got your emails.@",
                                    @"badge" : @(9),
                                    @"sound" : @"bingbong.aiff"},
                              @"acme 1" : @(42)};
    
    [self performVerifiedPublish:@"test"
                       onChannel:[self publishChannelName]
               mobilePushPayload:payload
                  storeInHistory:NO
                      compressed:NO
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                  }];
}

- (void)testPublishMobilePayloadToNillChannnel {
    
    NSDictionary *payload = @{@"aps" :
                                  @{@"alert" : @"You got your emails",
                                    @"badge" : @(9),
                                    @"sound" : @"bingbong.aiff"},
                              @"acme 1" : @(42)};
    
    [self performVerifiedPublish:@"test"
                       onChannel:nil
               mobilePushPayload:payload
                  storeInHistory:NO
                      compressed:NO
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                  }];
}

#pragma mark - Main flow

- (void)performVerifiedPublish:(id)message
                     onChannel:(NSString *)channel
             mobilePushPayload:(NSDictionary *)payload
                storeInHistory:(BOOL)storeInHistory
                    compressed:(BOOL)compressed
                withAssertions:(PNPublishCompletionBlock)verificationBlock {
    XCTestExpectation *networkExpectation = [self expectationWithDescription:@"network"];
    [self.client publish:message
               toChannel:channel
       mobilePushPayload:payload
          storeInHistory:storeInHistory
              compressed:compressed
          withCompletion:^(PNPublishStatus *status) {
              verificationBlock(status);
              [networkExpectation fulfill];
          }];
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

@end

File: Tests/iOS Tests-Swift/PNPublishWithMobilePayloadTests.swift
//
//  PNPublishWithMobilePayloadTests.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 10/20/15.
//
//

import UIKit
import XCTest

class PNPublishWithMobilePayloadTests: PNBasicClientTestCase {
    
    override func isRecording() -> Bool {
        return false
    }
    
    let channelName = "02290046-2F36-43DD-97F0-2F51D925451A"
    
    func testSimplePublishSimpleMobilePushPayload() {
        let payload = ["aps" :
            ["alert" : "You got your emails",
                "badge" : 9,
                "sound" : "bingbong.aiff"],
            "acme 1" : 42]
        
        self.performVerifiedPublish("test",
            onChannel: self.channelName,
            mobilePushPayload: payload,
            storeInHistory: true, compressed: true) { (status: PNPublishStatus!) -> Void in
                XCTAssertNotNil(status)
                XCTAssertTrue(status.category == .PNAcknowledgmentCategory)
                XCTAssertTrue(status.operation == .PublishOperation)
                XCTAssertEqual(status.statusCode, 200)
                XCTAssertFalse(status.error)
                XCTAssertEqual(status.data.information, "Sent")
                XCTAssertEqual(status.data.timetoken, 14355449397087311)
        }
    }
    
    func testSimplePublishNilMobilePushPayload() {
        
        self.performVerifiedPublish("test",
            onChannel: self.channelName,
            mobilePushPayload: nil,
            storeInHistory: true, compressed: true) { (status: PNPublishStatus!) -> Void in
                XCTAssertNotNil(status)
                XCTAssertTrue(status.category == .PNAcknowledgmentCategory)
                XCTAssertTrue(status.operation == .PublishOperation)
                XCTAssertEqual(status.statusCode, 200)
                XCTAssertFalse(status.error)
                XCTAssertEqual(status.data.information, "Sent")
                XCTAssertEqual(status.data.timetoken, 14355449396125368)
        }
    }
    
    func testPublishMobilePayloadNotStoreInHistory() {
        let payload = ["aps" :
            ["alert" : "You got your emails",
                "badge" : 9,
                "sound" : "bingbong.aiff"],
            "acme 1" : 42]
        
        self.performVerifiedPublish("test",
            onChannel: self.channelName,
            mobilePushPayload: payload,
            storeInHistory: false, compressed: true) { (status: PNPublishStatus!) -> Void in
                XCTAssertNotNil(status)
                XCTAssertTrue(status.category == .PNAcknowledgmentCategory)
                XCTAssertTrue(status.operation == .PublishOperation)
                XCTAssertEqual(status.statusCode, 200)
                XCTAssertFalse(status.error)
                XCTAssertEqual(status.data.information, "Sent")
                XCTAssertEqual(status.data.timetoken, 14355449393962620)
        }
    }

    func testPublishMobilePayloadNotStoreInHistoryNotCompressed() {
        let payload = ["aps" :
            ["alert" : "You got your emails",
                "badge" : 9,
                "sound" : "bingbong.aiff"],
            "acme 1" : 42]
        
        self.performVerifiedPublish("test",
            onChannel: self.channelName,
            mobilePushPayload: payload,
            storeInHistory: false, compressed: false) { (status: PNPublishStatus!) -> Void in
                XCTAssertNotNil(status)
                XCTAssertTrue(status.category == .PNAcknowledgmentCategory)
                XCTAssertTrue(status.operation == .PublishOperation)
                XCTAssertEqual(status.statusCode, 200)
                XCTAssertFalse(status.error)
                XCTAssertEqual(status.data.information, "Sent")
                // TODO: investigate me, changes every time
//                XCTAssertEqual(status.data.timetoken, 14453551891858438)
        }
    }
    
    func testPublishMobilePayloadToNillChannnel() {
        let payload = ["aps" :
            ["alert" : "You got your emails",
                "badge" : 9,
                "sound" : "bingbong.aiff"],
            "acme 1" : 42]
        
        self.performVerifiedPublish("test",
            onChannel: nil,
            mobilePushPayload: payload,
            storeInHistory: false, compressed: false) { (status: PNPublishStatus!) -> Void in
                XCTAssertNotNil(status)
                XCTAssertTrue(status.category == .PNBadRequestCategory)
                XCTAssertTrue(status.operation == .PublishOperation)
                XCTAssertEqual(status.statusCode, 400)
                XCTAssertTrue(status.error)
        }
    }
}

extension PNPublishWithMobilePayloadTests {
    func performVerifiedPublish(message: AnyObject,
        onChannel:String!,
        mobilePushPayload : [NSObject : AnyObject]!,
        storeInHistory: Bool,
        compressed: Bool,
        withAssertions: PNPublishCompletionBlock) {
            
        let networkExpectation = self.expectationWithDescription("network")
            
        self.client.publish(message, toChannel: onChannel, mobilePushPayload: mobilePushPayload, storeInHistory: storeInHistory, compressed: compressed) { (status: PNPublishStatus!) -> Void in
            withAssertions(status)
            networkExpectation.fulfill()
        }
            
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
}
